### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'url'

### ISSUE THREAD

**Issue by messense** [channelcat/sanic#483]
I'm using a custom error handler to send exception to sentry:

```
class SentryErrorHandler(ErrorHandler):
    def default(self, request, exception):
        if exception is None:
            return text('unknown error occurred')
        sentry.captureException()
        return super().default(request, exception)

app = Sanic(__name__, error_handler=SentryErrorHandler())
```

Got the exception:

```
2017-02-24 07:53:18,845 ERROR Process-22219 MainThread sanic.handlers.log:73    Traceback (most recent call last):
  File "/home/kpath/kpath/venv/lib/python3.6/site-packages/sanic/handlers.py", line 54, in response
    response = handler(request=request, exception=exception)
  File "app.py", line 486, in default
    sentry.captureException()
  File "/home/kpath/kpath/venv/lib/python3.6/site-packages/raven/base.py", line 799, in captureException
    'raven.events.Exception', exc_info=exc_info, **kwargs)
  File "/home/kpath/kpath/venv/lib/python3.6/site-packages/raven/base.py", line 615, in capture
    elif not self.should_capture(exc_info):
  File "/home/kpath/kpath/venv/lib/python3.6/site-packages/raven/base.py", line 803, in should_capture
    exc_name = '%s.%s' % (exc_type.__module__, exc_type.__name__)
AttributeError: 'NoneType' object has no attribute '__module__'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kpath/kpath/venv/lib/python3.6/site-packages/sanic/server.py", line 201, in write_error
    response = self.error_handler.response(self.request, exception)
  File "/home/kpath/kpath/venv/lib/python3.6/site-packages/sanic/handlers.py", line 61, in response
    handler.__name__, request.url, format_exc())
AttributeError: 'NoneType' object has no attribute 'url'

2017-02-24 07:53:18,845 ERROR Process-22219 MainThread sanic.server.bail_out:226   Writing error failed, connection closed AttributeError("'NoneType' object has no attribute 'url'",)
```

**Comment by r0fls** [comment]
So apparently the request is `None` for some reason..

**Comment by messense** [comment]
@r0fls Yes, but when will `request` be `None`?

**Comment by r0fls** [comment]
I'm assuming the connection didn't finish establishing or some other similar situation. It initializes to None in the protocol here: https://github.com/channelcat/sanic/blob/master/sanic/server.py#L69

**Comment by r0fls** [comment]
still a bug in the logging treatment either way

**Comment by seemethere** [comment]
The only places where this could happen would be:

* [Line 110](https://github.com/channelcat/sanic/blob/master/sanic/server.py#L110)
* [Line 122](https://github.com/channelcat/sanic/blob/master/sanic/server.py#L122)
* [Line 135](https://github.com/channelcat/sanic/blob/master/sanic/server.py#L135)
* [Line 143](https://github.com/channelcat/sanic/blob/master/sanic/server.py#L143)

Maybe it'd be better to try to check if the request was not finished being created? Or maybe instead of instantiating the request on message complete we instantiate when the data starts being sent so that we don't run into issues like this.

**Comment by messense** [comment]
I think at least it should feed the `request` object as much as it could when running into exceptions.

**Comment by zuoxiaoliangcoder** [comment]
if request is None,  the exception raise in server.py 116 to 123 line,  while httptools run the  data_received mehtod first.

**Comment by messense** [comment]
Fixed in #484
